pallets/foreign-investments/src/impls/invest.rs

```diff
@@ -0,0 +1,1087 @@
+	///   side with lower amounts; or both if the swap amounts are equal.
+	///
+	/// When we increase an investment, we normally have to swap it into pool
+	/// currency (`ActiveSwapIntoPoolCurrency`) before it can be invested
+	/// (`ActiveInvestmentOngoing`). However, if the current state includes
+	/// swapping back into pool currency (`ActiveSwapIntoForeignCurrency`) as
+	/// the result of a previous decrement, then we can minimize the amount
+	/// which needs to be swapped such that we always have **at most a single
+	/// active swap** which is the maximum of `pool_swap.amount` and
+	/// `foreign_swap.amount`. When we do this, we always need to bump the
+	/// investment amount as well as the `SwapIntoForeignDone` amount as a
+	/// result of immediately fulfilling the pool swap order up to the possible
+	/// amount.
+	///
+	/// Example:
+	/// * Say before my pre invest state has `foreign_done = 1000` and
+	/// `foreign_swap.amount = 500`. Now we  look at three scenarios in which we
+	/// increase below, exactly at and above the `foreign_swap.amount`:
+	/// * a) If we increase by 500, we can reduce the `foreign_swap.amount`
+	///   fully, which we denote by adding the 500 to the `foreign_done` amount.
+	///   Moreover, we can immediately invest the 500. The resulting state is
+	///   `(done_amount = 1500, investing = 500)`.
+	/// * b) If we increase by 400, we can reduce the `foreign_swap.amount` only
+	///   by 400 and increase both the `investing` as well as `foreign_done`
+	///   amount by that. The resulting state is
+	/// `(done_amount = 1400, foreign_swap.amount = 100, investing = 400)`.
+	/// * c) If we increase by 600, we can reduce the `foreign_swap.amount`
+	///   fully and need to add a swap into pool currency for 100. Moreover both
+	///   the `investing` as well as `foreign_done` amount can only be increased
+	///   by 500. The resulting state is
+	/// `(done_amount = 1500, pool_swap.amount = 100, investing = 500)`.
+	///
+	/// NOTES:
+	/// * We can never directly compare `swap.amount` and `invest_amount` with
+	///   `foreign_swap.amount` and `done_amount` if the currencies mismatch as
+	///   the former pair is denominated in pool currency and the latter one in
+	///   foreign currency.
+	/// * We can ignore handling all states which include `*SwapIntoForeignDone`
+	///   without `ActiveSwapIntoForeignCurrency*` as we consume the done amount
+	///   and transition in the post transition phase. To be safe and to not
+	///   make any unhandled assumptions, we throw `DispatchError::Other` for
+	///   these states though we need to make sure this can never occur!
+	fn handle_increase(
+		&self,
+		swap: Swap<T::Balance, T::CurrencyId>,
+	) -> Result<Self, DispatchError> {
+		if swap.currency_in == swap.currency_out {
+			return Self::handle_increase_non_foreign(self, swap);
+		}
+
+		match &self {
+			Self::NoState => Ok(Self::ActiveSwapIntoPoolCurrency { swap }),
```

Comment:
Here we are assuming the `swap` given in the `handle_increase()` parameter is a pool swap. What happens if the method is called with a foreign swap? Should that be handled and create an `ActiveSwapIntoForeignCurrency` instead? 

Actually, `ActiveSwapIntoForeignCurrency` is never created as an increase transition, only by a decrease transition, is that correct?

Reply:
> Actually, ActiveSwapIntoForeignCurrency is never created as an increase transition, only by a decrease transition, is that correct?

Yes! When we increase an investment, the incoming currency must be the `pool` one and the outgoing currency accepted as payment (i.e. either is pool currency or there exists a valid trading pair from `foreign` to `pool`). We ensure this in `[pallets/liquidity-pools/src/inbound.rs fn handle_increase_invest_order`](https://github.com/centrifuge/centrifuge-chain/blob/d46a36378796053e5c6adc4694ad4dc7a5ef3605/pallets/liquidity-pools/src/inbound.rs#L106-L108). 

The `InvestTransition` enum also has docs stating this:
https://github.com/centrifuge/centrifuge-chain/blob/d46a36378796053e5c6adc4694ad4dc7a5ef3605/pallets/foreign-investments/src/types.rs#L248-L249